File: lib/djv/App/TimelineActions.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the DJV project.

#include <djv/App/TimelineActions.h>

#include <djv/App/App.h>
#include <djv/App/MainWindow.h>

#include <tlRender/UI/TimelineWidget.h>

namespace djv
{
    namespace app
    {
        struct TimelineActions::Private
        {
            std::weak_ptr<MainWindow> mainWindow;

            std::shared_ptr<ftk::Observer<models::TimelineSettings> > settingsObserver;
        };

        void TimelineActions::_init(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<MainWindow>& mainWindow)
        {
            IActions::_init(context, app, "Timeline");
            FTK_P();

            p.mainWindow = mainWindow;

            // Register the commands.
            auto appWeak = std::weak_ptr<App>(app);
            _addCheckCommand(
                "Minimize",
                "Minimize the timeline.",
                [appWeak](const nlohmann::json& args)
                {
                    const bool value = args.at("value").get<bool>();
                    if (auto app = appWeak.lock())
                    {
                        auto settings = app->getSettingsModel()->getTimeline();
                        settings.minimize = value;
                        app->getSettingsModel()->setTimeline(settings);
                    }
                });

            _addCheckCommand(
                "FrameView",
                "Frame the timeline view.",
                [appWeak](const nlohmann::json& args)
                {
                    const bool value = args.at("value").get<bool>();
                    if (auto app = appWeak.lock())
                    {
                        auto settings = app->getSettingsModel()->getTimeline();
                        settings.frameView = value;
                        app->getSettingsModel()->setTimeline(settings);
                    }
                });

            _addCheckCommand(
                "ScrollBars",
                "Toggle the scroll bars.",
                [appWeak](const nlohmann::json& args)
                {
                    const bool value = args.at("value").get<bool>();
                    if (auto app = appWeak.lock())
                    {
                        auto settings = app->getSettingsModel()->getTimeline();
                        settings.scrollBars = value;
                        app->getSettingsModel()->setTimeline(settings);
                    }
                });

            _addCheckCommand(
                "AutoScroll",
                "Automatically scroll the timeline to the current frame.",
                [appWeak](const nlohmann::json& args)
                {
                    const bool value = args.at("value").get<bool>();
                    if (auto app = appWeak.lock())
                    {
                        auto settings = app->getSettingsModel()->getTimeline();
                        settings.autoScroll = value;
                        app->getSettingsModel()->setTimeline(settings);
                    }
                });

            _addCheckCommand(
                "StopOnScrub",
                "Stop playback when scrubbing the timeline.",
                [appWeak](const nlohmann::json& args)
                {
                    const bool value = args.at("value").get<bool>();
                    if (auto app = appWeak.lock())
                    {
                        auto settings = app->getSettingsModel()->getTimeline();
                        settings.stopOnScrub = value;
                        app->getSettingsModel()->setTimeline(settings);
                    }
                });

            _addCheckCommand(
                "Thumbnails",
                "Toggle timeline thumbnails.",
                [appWeak](const nlohmann::json& args)
                {
                    const bool value = args.at("value").get<bool>();
                    if (auto app = appWeak.lock())
                    {
                        auto settings = app->getSettingsModel()->getTimeline();
                        settings.thumbnails = value;
                        app->getSettingsModel()->setTimeline(settings);
                    }
                });

            _addCommand(
                "ThumbnailSizeSmall",
                "Small timeline thumbnails.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        auto settings = app->getSettingsModel()->getTimeline();
                        settings.thumbnailSize = models::TimelineThumbnailSize::Small;
                        app->getSettingsModel()->setTimeline(settings);
                    }
                });

            _addCommand(
                "ThumbnailSizeMedium",
                "Medium timeline thumbnails.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        auto settings = app->getSettingsModel()->getTimeline();
                        settings.thumbnailSize = models::TimelineThumbnailSize::Medium;
                        app->getSettingsModel()->setTimeline(settings);
                    }
                });

            _addCommand(
                "ThumbnailSizeLarge",
                "Large timeline thumbnails.",
                [appWeak](const nlohmann::json&)
                {
                    if (auto app = appWeak.lock())
                    {
                        auto settings = app->getSettingsModel()->getTimeline();
                        settings.thumbnailSize = models::TimelineThumbnailSize::Large;
                        app->getSettingsModel()->setTimeline(settings);
                    }
                });

            // Create the actions.
            _actions["Minimize"] = ftk::Action::create(
                "Minimize",
                _checkCommand("Minimize"));
            _actions["FrameView"] = ftk::Action::create(
                "Frame View",
                _checkCommand("FrameView"));
            _actions["ScrollBars"] = ftk::Action::create(
                "Scroll Bars",
                _checkCommand("ScrollBars"));
            _actions["AutoScroll"] = ftk::Action::create(
                "Auto Scroll",
                _checkCommand("AutoScroll"));
            _actions["StopOnScrub"] = ftk::Action::create(
                "Stop Playback When Scrubbing",
                _checkCommand("StopOnScrub"));
            _actions["Thumbnails"] = ftk::Action::create(
                "Thumbnails",
                _checkCommand("Thumbnails"));
            _actions["ThumbnailSizeSmall"] = ftk::Action::create(
                "Small",
                _command("ThumbnailSizeSmall"));
            _actions["ThumbnailSizeMedium"] = ftk::Action::create(
                "Medium",
                _command("ThumbnailSizeMedium"));
            _actions["ThumbnailSizeLarge"] = ftk::Action::create(
                "Large",
                _command("ThumbnailSizeLarge"));

            _shortcutsUpdate(app->getSettingsModel()->getShortcuts());

            p.settingsObserver = ftk::Observer<models::TimelineSettings>::create(
                app->getSettingsModel()->observeTimeline(),
                [this](const models::TimelineSettings& value)
                {
                    _actions["Minimize"]->setChecked(value.minimize);
                    _actions["FrameView"]->setChecked(value.frameView);
                    _actions["ScrollBars"]->setChecked(value.scrollBars);
                    _actions["AutoScroll"]->setChecked(value.autoScroll);
                    _actions["StopOnScrub"]->setChecked(value.stopOnScrub);
                    _actions["Thumbnails"]->setChecked(value.thumbnails);
                    _actions["ThumbnailSizeSmall"]->setChecked(models::TimelineThumbnailSize::Small == value.thumbnailSize);
                    _actions["ThumbnailSizeMedium"]->setChecked(models::TimelineThumbnailSize::Medium == value.thumbnailSize);
                    _actions["ThumbnailSizeLarge"]->setChecked(models::TimelineThumbnailSize::Large == value.thumbnailSize);
                });
        }

        TimelineActions::TimelineActions() :
            _p(new Private)
        {}

        TimelineActions::~TimelineActions()
        {}

        std::shared_ptr<TimelineActions> TimelineActions::create(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::shared_ptr<MainWindow>& mainWindow)
        {
            auto out = std::shared_ptr<TimelineActions>(new TimelineActions);
            out->_init(context, app, mainWindow);
            return out;
        }
    }
}

File: examples/python/TimelineActions.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the DJV project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl
import djvPy as djv

import IActions
import Util

import weakref

class Actions(IActions.IActions):
    """
    This class provides the timeline actions.
    """
    def __init__(self, context, app):
        IActions.IActions.__init__(self, context, app, "Timeline")

        self._app = weakref.ref(app)

        # Register the commands.
        checks = [
            ("Minimize", "Minimize", "minimize", "Minimize the timeline."),
            ("FrameView", "Frame View", "frameView", "Frame the timeline view."),
            ("ScrollBars", "Scroll Bars", "scrollBars", "Toggle the scroll bars."),
            ("AutoScroll", "Auto Scroll", "autoScroll",
             "Automatically scroll the timeline to the current frame."),
            ("StopOnScrub", "Stop When Scrubbing", "stopOnScrub",
             "Stop playback when scrubbing the timeline."),
            ("TrackMedia", "Track Media", "trackMedia",
             "Toggle the timeline video thumbnails and audio waveforms."),
            ("Thumbnails", "Thumbnails", "thumbnails",
             "Toggle the timeline video thumbnails."),
            ("Waveforms", "Waveforms", "waveforms",
             "Toggle the timeline audio waveforms."),
        ]
        for name, text, field, doc in checks:
            self._addCheckCommand(
                name,
                doc,
                lambda value, captured = field, \
                    f = Util.weak(self._setField): f(captured, value))

        sizes = [
            ("ThumbnailSizeSmall", "Small", "thumbnailSize",
             djv.models.TimelineThumbnailSize.Small, "Small timeline thumbnails."),
            ("ThumbnailSizeMedium", "Medium", "thumbnailSize",
             djv.models.TimelineThumbnailSize.Medium, "Medium timeline thumbnails."),
            ("ThumbnailSizeLarge", "Large", "thumbnailSize",
             djv.models.TimelineThumbnailSize.Large, "Large timeline thumbnails."),
            ("WaveformSizeSmall", "Small", "waveformSize",
             djv.models.TimelineThumbnailSize.Small, "Small timeline audio waveforms."),
            ("WaveformSizeMedium", "Medium", "waveformSize",
             djv.models.TimelineThumbnailSize.Medium, "Medium timeline audio waveforms."),
            ("WaveformSizeLarge", "Large", "waveformSize",
             djv.models.TimelineThumbnailSize.Large, "Large timeline audio waveforms."),
        ]
        for name, text, field, size, doc in sizes:
            self._addCommand(
                name,
                doc,
                lambda args, capturedField = field, capturedSize = size, \
                    f = Util.weak(self._setField): f(capturedField, capturedSize))

        # Create the actions.
        for name, text, field, doc in checks:
            self.actions[name] = ftk.Action(
                text,
                checkedCallback = self._checkCommand(name))
        for name, text, field, size, doc in sizes:
            self.actions[name] = ftk.Action(
                text,
                self._command(name))

        # Register the shortcuts.
        self._addShortcut("Minimize")
        self._addShortcut("FrameView")
        self._addShortcut("ScrollBars")
        self._addShortcut("AutoScroll")
        self._addShortcut("StopOnScrub", "Stop On Scrub")
        self._addShortcut("Thumbnails")
        self._addShortcut("ThumbnailSizeSmall", "Small Video Thumbnails")
        self._addShortcut("ThumbnailSizeMedium", "Medium Video Thumbnails")
        self._addShortcut("ThumbnailSizeLarge", "Large Video Thumbnails")
        self._addShortcut("Waveforms")
        self._addShortcut("TrackMedia", ftk.Key.C)
        self._addShortcut("WaveformSizeSmall", "Small Audio Waveforms")
        self._addShortcut("WaveformSizeMedium", "Medium Audio Waveforms")
        self._addShortcut("WaveformSizeLarge", "Large Audio Waveforms")

        self._shortcutsUpdate(self._settingsModel.shortcuts)

        selfWeak = weakref.ref(self)
        self._settingsObserver = djv.models.TimelineSettingsObserver(
            app.getSettingsModel().observeTimeline,
            lambda value: selfWeak()._settingsUpdate(value))

    def _setField(self, field, value):
        model = self._app().getSettingsModel()
        settings = model.timeline
        setattr(settings, field, value)
        model.timeline = settings

    def _settingsUpdate(self, settings):
        self.actions["Minimize"].checked = settings.minimize
        self.actions["FrameView"].checked = settings.frameView
        self.actions["ScrollBars"].checked = settings.scrollBars
        self.actions["AutoScroll"].checked = settings.autoScroll
        self.actions["StopOnScrub"].checked = settings.stopOnScrub
        self.actions["TrackMedia"].checked = settings.trackMedia
        self.actions["Thumbnails"].checked = settings.thumbnails
        self.actions["Waveforms"].checked = settings.waveforms
        Small = djv.models.TimelineThumbnailSize.Small
        Medium = djv.models.TimelineThumbnailSize.Medium
        Large = djv.models.TimelineThumbnailSize.Large
        self.actions["ThumbnailSizeSmall"].checked = Small == settings.thumbnailSize
        self.actions["ThumbnailSizeMedium"].checked = Medium == settings.thumbnailSize
        self.actions["ThumbnailSizeLarge"].checked = Large == settings.thumbnailSize
        self.actions["WaveformSizeSmall"].checked = Small == settings.waveformSize
        self.actions["WaveformSizeMedium"].checked = Medium == settings.waveformSize
        self.actions["WaveformSizeLarge"].checked = Large == settings.waveformSize
